### DISCUSSION TITLE
Expose allowed HTTP methods from endpoints and pages

### DISCUSSION THREAD

**Discussion by natemoo-re** [withastro/roadmap#723] [Proposal]
### Body

# Summary

It would be helpful for Astro to expose the accepted HTTP methods for any routes. Astro servers should probably return a [`405 Method Not Allowed`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/405) status for disallowed methods.

# Background & Motivation

Dax (SST) mentioned in Discord that Astro routes being able to accept _any_ method causes some issues for SST deployment.

> so in astro there’s nowhere people have to explicitly say if a route accepts a GET or a POST right?
> this is convenient but it creates problems for deployment adapters - that info is important to keep certain things performant
> it’s something we’d like to know at build time because it effects how we route things in the cdn

# Goals

- Improve the performance of Astro sites by exposing information necessary for adapters to optimize their deployments.
  - Would this be in the route manifest in the `astro:build:done` hook? We've wanted to redesign that interface, so maybe this is a good opportunity to do so.
- The integration API should expose acceptable HTTP methods for each route
  - Determining these for endpoints is straightforward. If GET and POST are exported, those are the accepted methods. If a `default` handler is exported, it could accept any method.
  - Astro pages are tricky. We might need a new API for users to declare what methods are supported? Static analysis could be another (more error prone) option.

**Comment by mb21** [comment]
> Astro pages are tricky. We might need a new API for users to declare what methods are supported?

Yes, to be consistent with endpoints' syntax, it should look something like:

```jsx
export async function GET(context) {
  return <h1>My page</h1>
}
```

(The current syntax could still continue to be supported, and would imply that it's a GET-only HTML endpoint.)

Does the above syntax look familiar? 😜 Okay, I admit it: this is a shameless plug for https://github.com/withastro/roadmap/discussions/716
